File: jdave-examples/src/test/jdave/examples/StackSpec.java
/*
 * Copyright 2006 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jdave.examples;

import org.junit.runner.RunWith;

import jdave.Block;
import jdave.Group;
import jdave.Specification;
import jdave.junit4.JDaveRunner;

/**
 * @author Joni Freeman
 * @author Pekka Enberg
 */
@RunWith(JDaveRunner.class)
@Group("basic")
public class StackSpec extends Specification<Stack<?>> {
    public class EmptyStack {
        private Stack<String> stack;

        public Stack<String> create() {
            stack = new Stack<String>();
            return stack;
        }

        public void isEmpty() {
            specify(stack, should.be.empty());
        }

        public void isNoLongerEmptyAfterPush() {
            stack.push("anything");
            specify(stack, should.not().be.empty());
        }
    }

    public class FullStack {
        private Stack<Integer> stack;

        public Stack<Integer> create() {
            stack = new Stack<Integer>(10);
            for (int i = 0; i < 10; i++) {
                stack.push(i);
            }
            return stack;
        }

        public void isFull() {
            specify(stack, should.be.full());
        }

        public void complainsOnPush() {
            specify(new Block() {
                public void run() throws Throwable {
                    stack.push(100);
                }
            }, should.raise(StackOverflowException.class));
        }
        
        public void containsAllItems() {
            for (int i = 0; i < 10; i++) {
                specify(stack, contains(i));
            }
        }
        
        public void doesNotContainRemovedItem() {
            stack.remove(3);
            specify(stack, does.not().contain(3));
        }
        
        public void containsAllButRemovedItem() {
            stack.remove(3);
            specify(stack, containsExactly(0, 1, 2, 4, 5, 6, 7, 8, 9));
        }
    }

    public class StackWhichIsNeitherEmptyNorFull {
        private Stack<Integer> stack;

        public Stack<Integer> create() {
            stack = new Stack<Integer>();
            for (int i = 0; i < 10; i++) {
                stack.push(i);
            }
            return stack;
        }

        public void addsToTheTopWhenSentPush() {
            stack.push(100);
            specify(stack.peek(), should.equal(100));
        }
    }
}

File: jdave-scala/src/test/jdave/scala/StackSpec.scala
/*
 * Copyright 2006 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jdave.scala;

import org.junit.runner.RunWith;
import jdave.junit4.JDaveRunner;

@RunWith(classOf[JDaveRunner])
class StackSpec extends Specification[Stack[Int]] {
  class EmptyStack {
    val stack = new Stack[Int]
    def create = stack
    
    def isEmpty = §(stack, is empty)

    def isNoLongerEmptyAfterPush {
      stack.push(1234)
     // §(stack, must not() be empty)
    }
  }
  
  class FullStack {
    val stack = new Stack[Int](10)

    def create = {
      (0 to 9).foreach(stack.push)
      stack
    }

    def isFull = §(stack, is full)
    def complainsOnPush = §({ stack.push(100) }, must raise classOf[StackOverflowException])
    def containsAllItems = (0 to 9).foreach(i => §(stack, contains(i)))
    
    def doesNotContainRemovedItem {
      stack.pop()
      §(stack, does not() contain 9)
    }
    
    def containsAllButRemovedItem {
      stack.pop()
      §(stack, containsExactly(List(0, 1, 2, 3, 4, 5, 6, 7, 8)))
    }
  }

  class StackWhichIsNeitherEmptyNorFull {
    val stack = new Stack[Int]

    def create = {
      (0 to 9).foreach(stack.push)
      stack
    }

    def addsToTheTopWhenSentPush {
      stack.push(100)
      §(stack.peek, must equal 100)
    }
  }
}
